driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -102,120 +84,101 @@ class AsyncCommandBatchCursor<T> implements AsyncAggregateResponseBatchCursor<T>
-            this.connectionSource = notNull("connectionSource", connectionSource).retain();
-            assertNotNull(connection);
-            if (limitReached()) {
-                killCursor(connection);
-            } else {
-                if (connectionSource.getServerDescription().getType() == ServerType.LOAD_BALANCER) {
-                    this.pinnedConnection = connection.retain();
-                    this.pinnedConnection.markAsPinned(Connection.PinningMode.CURSOR);
-                }
-            }
+        this.maxWireVersion = connectionDescription.getMaxWireVersion();
+        this.firstBatchEmpty = commandCursorResult.getResults().isEmpty();
+
+        AsyncConnection connectionToPin = null;
+        boolean releaseServerAndResources = false;
+        if (limitReached()) {
+            releaseServerAndResources = true;
+        } else if (connectionDescription.getServerType() == ServerType.LOAD_BALANCER) {
+            connectionToPin = connection;
         }
-        logQueryResult(initialCommandCursorResult);
-    }
-
-    /**
-     * {@inheritDoc}
-     * <p>
-     * From the perspective of the code external to this class, this method is idempotent as required by its specification.
-     * However, if this method sets {@link #isClosePending},
-     * then it must be called by {@code this} again to release resources.
-     * This behavior does not violate externally observable idempotence because this method is allowed to release resources "eventually".
-     */
-    @Override
-    public void close() {
-        boolean doClose = withLock(lock, () -> {
-            if (isOperationInProgress) {
-                isClosePending = true;
-                return false;
-            } else if (!isClosed) {
-                isClosed = true;
-                isClosePending = false;
-                return true;
-            }
-            return false;
-        });
 
-        if (doClose) {
-            killCursorOnClose();
+        resourceManager = new ResourceManager(namespace, connectionSource, connectionToPin, commandCursorResult.getServerCursor());
+        if (releaseServerAndResources) {
+            resourceManager.releaseServerAndClientResources(connection);
         }
     }
 
     @Override
     public void next(final SingleResultCallback<List<T>> callback) {
         if (isClosed()) {
-            callback.onResult(null, new MongoException("next() called after the cursor was closed."));
-        } else if (initialCommandCursorResult != null && !firstBatchEmpty) {
-            // May be empty for a tailable cursor
-            List<T> results = initialCommandCursorResult.getResults();
-            initialCommandCursorResult = null;
-            if (getServerCursor() == null) {
-                close();
+            callback.onResult(null, new MongoException(MESSAGE_IF_CLOSED_AS_CURSOR));
+            return;
+        }
+
+
+
+        resourceManager.execute((AsyncCallbackSupplier<List<T>>) funcCallback -> {
```

Comment:
`funcCallback` is a bit ambiguous.  How about `resourceManagerCallback`?

Reply:
`funcCallback` is the existing convention in the codebase where `AsyncCallbackSupplier<>` is used and a variable called `callback` is in scope.